### PR TITLE
Feature/percent ingest

### PR DESCRIPTION
**Pull request by chdietz**
Parse percentile CSV file for Percentiles
Percentile and PercentileScore are parsed from CSV and align with warehouse tables 
  percentile
  percentile_score

The CSV columns that map to percentile do not have to be repeated for adjacent percentile_score columns except the columns that identify the percentile table that is currently `assessment_id`.

The columns are:
`assessment_id, count, mean, standard_deviation, percent, score`

### CODE REVIEW

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+import java.util.List;
+
+import static org.opentestsystem.rdw.common.model.ImportStatus.BAD_DATA;
+import static org.opentestsystem.rdw.common.model.ImportStatus.BAD_FORMAT;
+import static org.opentestsystem.rdw.ingest.common.util.ParserHelper.toDouble;
+import static org.opentestsystem.rdw.ingest.common.util.ParserHelper.toInteger;
+import static org.springframework.util.StringUtils.hasText;
+
+@Service
+class DefaultPercentileParser implements PercentileParser {
```
Comment (chdietz) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java]: based on how we parse assessment packeges

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                if (!percentileReader.isSamePercentile(percentileRecord)) {
+                    percentiles.add(percentileReader.safeBuild(percentilesErrorCollector));
+                    percentileReader = new PercentileReader(percentileRecord);
+                } else {
+                    percentileReader.addPercentileScore(percentileRecord);
+                }
+            }
+            percentiles.add(percentileReader.safeBuild(percentilesErrorCollector));
+
+            return percentilesErrorCollector.isEmpty() ? percentiles : new ArrayList<>();
```
Comment (malaffoon) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java]: Shouldn't this be throwing or something if there are any errors?

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/PercentileParser.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import org.opentestsystem.rdw.ingest.processor.model.Percentile;
+
+import java.util.List;
+
+/**
+ * Parses percentile payloads
+ */
+public interface PercentileParser {
+
+    List<Percentile> parse(byte[] payload, DataElementErrorCollector elementErrorCollector);
```
Comment (malaffoon) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/PercentileParser.java]: javadocs

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/PercentileParserTestIT.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+@RunWith(SpringRunner.class)
+@SpringBootTest
+@Transactional
+@CachingTest
+@ActiveProfiles("test")
+@ContextConfiguration(classes = {PackageProcessorApplication.class})
+@SqlGroup({
+        @Sql(executionPhase = BEFORE_TEST_METHOD, scripts = {"classpath:/PreloadAsmt.sql"})
+})
+public class PercentileParserTestIT {
```
Comment (malaffoon) [inline package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/PercentileParserTestIT.java]: Seems like you could pretty easily make this a `Test` instead of a `IT`.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/PercentileScore.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
+package org.opentestsystem.rdw.ingest.processor.model;
+
+public class PercentileScore {
+    private Integer percent;
```
[the remark this answers is not in the document]
Comment (chdietz) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/PercentileScore.java]: I guess it is most likely in the upper range, so if I make it a double I will want to change the persistence from tinyint to float.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+    private final AssessmentPackageRepository assessmentPackageRepository;
+
+    @Autowired
+    public DefaultPercentileParser(final AssessmentPackageRepository assessmentPackageRepository) {
+        this.assessmentPackageRepository = assessmentPackageRepository;
+    }
+
+    public List<Percentile> parse(final byte[] percentilePayload, final DataElementErrorCollector percentilesErrorCollector) {
+        try (final Reader reader = new InputStreamReader(new ByteArrayInputStream(percentilePayload), Charset.forName("utf8"))) {
```
[the remark this answers is not in the document]
Comment (chdietz) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java]: Did you pick this up from recent issues with upload student group CSV ?

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/PercentileParser.java (hunk 3 above)
Comment (chdietz) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/PercentileParser.java]: noted ;)

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/PercentileParserTestIT.java (hunk 4 above)
Comment (chdietz) [reply package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/PercentileParserTestIT.java]: Fair point I will add some unit tests and I can mock the repo. I do not see an issue having at least one IT.  I found out something i did not know that the findOneByNaturalId joins on asmt_score.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java (hunk 2 above)
Comment (chdietz) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java]: Yes, another copy what was there.  I'll look at this when I do the processor, we should probably have both work the same way.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Percentile.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Percentile {
+    private Integer id;
+    private Integer assessmentId;
+    private Integer count;
+    private Double mean;
+    private Double standardDeviation;
```
[the remark this answers is not in the document]
Comment (chdietz) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Percentile.java]: I was thinking count and standardDeviation could be optional

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            if (!recordIterator.hasNext()) return percentiles;
+
+            PercentileReader percentileReader = new PercentileReader(recordIterator.next());
+
+            while (recordIterator.hasNext()) {
+                final CSVRecord percentileRecord = recordIterator.next();
+
+                if (!percentileReader.isSamePercentile(percentileRecord)) {
+                    percentiles.add(percentileReader.safeBuild(percentilesErrorCollector));
+                    percentileReader = new PercentileReader(percentileRecord);
```
[the remark this answers is not in the document]
Comment (chdietz) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java]: @malaffoon your concern is un necessary object creation I assume?
I start with the assessment parser and found that pattern to be a bit hard to follow,
so this was my first cut.

[the remark this answers is not in the document]
Comment (malaffoon) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java]: Yes. I won't fight you for it but i prefer to avoid obvious excessive object creation.

Also, I really don't like naked `new` ... and you have one in a loop!  ;)

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/PercentileScore.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (chdietz) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/PercentileScore.java]: I will address this but update the schema first
